src/query/service/src/pipelines/processors/transforms/transform_sort_spill.rs

```diff
@@ -0,0 +1,745 @@
+        match &self.state {
+            State::Spill => {
+                let block = self.input_data.take().unwrap();
+                self.spill(block).await?;
+            }
+            State::Merging => {
+                let block = self.input_data.take();
+                self.merge_sort(block).await?;
+            }
+            State::MergeFinished => {
+                debug_assert_eq!(self.unmerged_blocks.len(), 1);
+                // TODO: pass the spilled locations to next processor directly.
+                // The next processor will read and process the spilled files.
+                if let Some(file) = self.unmerged_blocks[0].pop_front() {
+                    let block = self.spiller.read_spilled(&file).await?;
+                    self.output_data = Some(block);
+                } else {
+                    self.state = State::Finish;
+                }
+            }
+            _ => unreachable!(),
+        }
+        Ok(())
+    }
+}
+
+impl<R> TransformSortSpill<R>
+where R: Rows + Sync + Send + 'static
+{
+    pub fn create(
+        input: Arc<InputPort>,
+        output: Arc<OutputPort>,
+        schema: DataSchemaRef,
+        sort_desc: Arc<Vec<SortColumnDescription>>,
+        spiller: Spiller,
+    ) -> Self {
+        // TODO(spill): `num_merge` and `batch_size` should be determined by the memory usage.
+        Self {
+            input,
+            output,
+            schema,
+            input_data: None,
+            output_data: None,
+            spiller,
+            state: State::Init,
+            num_merge: 2,
+            unmerged_blocks: VecDeque::new(),
+            batch_size: 65535,
+            sort_desc,
+            _r: PhantomData,
+        }
+    }
+
+    async fn spill(&mut self, block: DataBlock) -> Result<()> {
+        let location = self.spiller.spill_block(block).await?;
+        self.unmerged_blocks.push_back(vec![location].into());
+        Ok(())
+    }
+
+    /// Do an external merge sort until there is only one sorted stream.
+    /// If `block` is not [None], we need to merge it with spilled files.
+    async fn merge_sort(&mut self, mut block: Option<DataBlock>) -> Result<()> {
+        while (self.unmerged_blocks.len() + block.is_some() as usize) > 1 {
+            let b = block.take();
+            self.merge_sort_one_round(b).await?;
+        }
+        self.state = State::MergeFinished;
+        Ok(())
+    }
+
+    /// Merge certain number of sorted streams to one sorted stream.
+    async fn merge_sort_one_round(&mut self, block: Option<DataBlock>) -> Result<()> {
+        let mut num_streams = self.unmerged_blocks.len() + block.is_some() as usize;
+        debug_assert!(num_streams > 1);
+        num_streams = num_streams.min(self.num_merge);
+
+        let mut streams = Vec::with_capacity(num_streams);
+        if let Some(block) = block {
+            streams.push(BlockStream::Block(Some(block)));
+            num_streams -= 1;
```

Comment:
why ?

Reply:
To guarantee we only merge `num_merge` streams.